app/com/linkedin/drelephant/spark/heuristics/UnifiedMemoryHeuristic.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2016 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.linkedin.drelephant.spark.heuristics
+
+import com.linkedin.drelephant.analysis._
+import com.linkedin.drelephant.configurations.heuristic.HeuristicConfigurationData
+import com.linkedin.drelephant.spark.data.SparkApplicationData
+import com.linkedin.drelephant.spark.fetchers.statusapiv1.ExecutorSummary
+import scala.collection.JavaConverters
+
+
+/**
+  * A heuristic based on peak unified memory for the spark executors
+  *
+  * This heuristic reports the fraction of memory used/ memory allocated for execution and if the fraction can be reduced. Also, it checks for the skew in peak unified memory and reports if the skew is too much.
+  */
+
+
+class UnifiedMemoryHeuristic(private val heuristicConfigurationData: HeuristicConfigurationData)
+  extends Heuristic[SparkApplicationData] {
+
+  import UnifiedMemoryHeuristic._
+  import JavaConverters._
+
+  override def getHeuristicConfData(): HeuristicConfigurationData = heuristicConfigurationData
+
+  override def apply(data: SparkApplicationData): HeuristicResult = {
+    val evaluator = new Evaluator(this, data)
+
+    var resultDetails = Seq(
+      new HeuristicResultDetails("Max peak unified memory", evaluator.maxUnifiedMemory.toString),
+      new HeuristicResultDetails("Mean peak unified memory", evaluator.meanUnifiedMemory.toString)
+    )
+
+    if (evaluator.severityPeak.getValue > Severity.LOW.getValue) {
+      resultDetails = resultDetails :+ new HeuristicResultDetails("Note", "The value of peak unified memory is very low, we recommend to increase the value of spark.memory.fraction")
+    }
+    if (evaluator.severitySkew.getValue > Severity.LOW.getValue) {
+      resultDetails = resultDetails :+ new HeuristicResultDetails("Note", "There is a imbalance in the amount of work (and data) for tasks, please see into this so that in can be more balanced")
+    }
+    val result = new HeuristicResult(
+      heuristicConfigurationData.getClassName,
+      heuristicConfigurationData.getHeuristicName,
+      evaluator.severity,
+      0,
+      resultDetails.asJava
+    )
+    result
+  }
+
+}
+
+object UnifiedMemoryHeuristic {
+
+  val JVM_USED_MEMORY = "jvmUsedMemory"
```

Comment:
This should be replaced by vals for storage memory and execution memory.

Reply:
I didn't get this, could you please be more specific?

Reply:
Unified memory is the sum of storage and execution memory. This is different from JVM used memory.
Storage memory is the amount of memory used by Spark for storing RDDS, broadcast variables, etc.
Execution memory is the amount of memory used for execution (shuffle, sort, etc.).
For more recent versions of Spark, execution and storage memory share a unified memory region, and are able to borrow from each other, if one still has extra capacity. Unified memory is thus the sum of storage and execution memory.
JVM memory is the JVM memory used by the Spark application.

Reply:
Yupp I understand that. Now peakUnifiedMemory contains the following key value pairs: 

“peakUnifiedMemory” : {
    “jvmUsedMemory” : <val>,
    “executionMemory”: <val>,
    “storageMemory”: <val>,
    “time”: <val>,
    “activeStages”: [<stage-id>, ...]
  }

 so for the purpose of this heuristic, to calculate the peak unified memory, do you mean that I should take sum of executionMemory and storageMemory instead of jvmUsedMemory?

Reply:
Yes, please use the sum of executionMemory and storageMemory for unifiedMemory.

Reply:
Sure.. Thank you! :)